### ISSUE TITLE
AST nodes with loc and/or range fields.

### ISSUE THREAD

**Issue by knightedcodemonkey** [oxc-project/oxc#10307]
It would be great to have AST nodes with a `loc` and/or `range` fields.

Here's some context: https://github.com/oxc-project/oxc/issues/10139#issuecomment-2783567533

Possibly dependent upon https://github.com/oxc-project/oxc/issues/9705 being completed first.

**Comment by alexander-akait** [comment]
@overlookmotel yeah, I understand this, I wanted to do a fairly powerful testing through many webpack tests and places, but without these things, especially without `loc`, it is not possible, will be wait updates

**Comment by overlookmotel** [comment]
More testing would be very welcome! Thank you.

On TS-ESLint alignment, in 2 weeks, we've gone from 1% of tests passing to 75%. So hopefully it won't be too long until #9705 is complete, and then we can look at this. By all means please get involved in the effort on #9705 if you'd like to. More hands will get us to the finish line faster.

**Comment by overlookmotel** [comment]
#11728 will cover adding `range`.

### Next steps - `range`

After it's merged, we'll need to:

* [ ] Adapt ESTree conformance testing to check `range` field is present and correct in all AST nodes (it'd be easy to forget it in some of the types which have custom `ESTree` implementations).
* [x] Support `range` field in raw transfer too.
* [x] Update TS type definitions to include `range` field.

### `loc`

Unfortunately I think supporting `loc` is more tricky, due to UTF-8 / UTF-16 offsets.

Here are the steps I think we'd need to take to support it:

* [x] Move `start`, `end` and `range` to be the last properties in each node object.
* [x] Adapt `acorn-test262` to do the same, so snapshots in conformance test have these properties in right order.
* [ ] Remove the extra AST pass converting offsets UTF-8 -> UTF-16. Convert during serialization instead.
* [ ] Do the same in raw transfer.
* [ ] Alter the [offsets translation table](https://github.com/oxc-project/oxc/blob/80c87d47b33cfb2afbb802af9c9d8e511275859f/crates/oxc_ast_visit/src/utf8_to_utf16.rs) to also represent line breaks.
* [ ] Use the new translation table to also convert UTF-8 offset -> line + column during serialization.

The last 4 steps are non-trivial.

I have only a vague idea of what a new translation table would look like. Ideally we probably want a single translation table that performs both UTF-8 to UTF-16 offset conversion, *and* UTF-8 offset to line / column conversion - rather than 2 separate tables.

### Why do location properties have to be last?

Conversion is far more efficient if you convert offsets in strictly ascending order (i.e. source code order).

The current converter ([generated code](https://github.com/oxc-project/oxc/blob/80c87d47b33cfb2afbb802af9c9d8e511275859f/crates/oxc_ast_visit/src/generated/utf8_to_utf16_converter.rs), [codegen](https://github.com/oxc-project/oxc/blob/80c87d47b33cfb2afbb802af9c9d8e511275859f/tasks/ast_tools/src/generators/utf8_to_utf16.rs)) works around various kinks in the AST to ensure this is the case.

But generally speaking, you need to convert the `start` offset of parent first, then all that nodes's children, and then finally the parent's `end`.

e.g. for `x + y`, order of conversion is:

* `start` of `BinaryExpression`.
* `start` of `Identifier` `x`.
* `end` of `Identifier` `x`.
* `start` of `Identifier` `y`.
* `end` of `Identifier` `y`.
* `end` of `BinaryExpression`.

Because JSON is constructed as a stream, `end` property must be the last property you output, because you don't know what it is until after all the children have been serialized.

```json
{
  "type": "BinaryExpression",
  "left": {
    "type": "Identifier",
    "name": "x",
    "start": 0,
    "end": 1,
    "range": [0, 1]
  },
  "operator": "+",
  "right": {
    "type": "Identifier",
    "name": "y",
    "start": 4,
    "end": 5,
    "range": [4, 5]
  },
  "start": 0,
  "end": 5,
  "range": [0, 5]
}
```

**Comment by overlookmotel** [comment]
#11728 is merged. Thank you @bacarybruno!

However, that PR was a bit incomplete, and we need some follow-up to "finish the job":

* [x] Add `range` field to all the custom hand-written `ESTree` impls in [`oxc_ast/src/serialize` directory](https://github.com/oxc-project/oxc/tree/9a2548ae9dd9d218a60673db87ff139316d0525c/crates/oxc_ast/src/serialize).
* [x] Add `range` field to TS types. This can be done by altering [the codegen which generates the type defs](https://github.com/oxc-project/oxc/blob/9a2548ae9dd9d218a60673db87ff139316d0525c/tasks/ast_tools/src/generators/typescript.rs).
* [x] Document the `range` option in [README](https://github.com/oxc-project/oxc/blob/9a2548ae9dd9d218a60673db87ff139316d0525c/napi/parser/README.md) for `oxc-parser` package.

**Comment by bacarybruno** [comment]
I plan to work on the remaining tasks for the `range` field (here is the first one https://github.com/oxc-project/oxc/pull/11890). Then, if it's okay with you, I can try to take simple tasks for the `loc` field if it can help.

EDIT: just saw that you are taking care of the other tasks 👍

**Comment by overlookmotel** [comment]
I've not attempted the 2 items listed in "next steps - range" in https://github.com/oxc-project/oxc/issues/10307#issuecomment-2976882861. Both of these are tricky, and I'm not going to be able to get to them immediately. I doubt, unfortunately, that they'd be something a contributor who doesn't know all the ins and out of raw transfer could work on.

We need those 2 ticked off before we tackle adding `loc` field - which will be trickier still. So I'm afraid we're blocked for now. I'll come back to it as soon as I'm able.

**Comment by overlookmotel** [comment]
Had some thoughts on the UTF-8 -> UTF-16 translation table. Writing them down now before I forget...

I think we need 2 separate tables:

```rs
struct Utf8ToUtf16 {
    lines: Vec<Translation>,
    offsets: Vec<Translation>,
}

struct Translation {
    utf8_offset: u32,
    utf16_difference: u32,
}
```

* `lines` contains an entry for each line in the source.
  * `utf8_offset` is the byte offset of the start of line.
  * `utf16_difference` is the UTF-8 / UTF-16 character count difference at start of the line.
  * Index is the line number.
* `offsets` contains an entry for each non-ASCII character in the source.
  * `utf8_offset` is the byte offset of start of that character + 1.
  * `utf16_difference` is the UTF-8 / UTF-16 character count difference after this character.

Conversion from UTF-8 offset:

* To UTF-16 offset: Search `offsets` for last entry where `offset.utf8_offset < utf8_offset`. `utf16_offset = utf8_offset - offset.utf16_difference` (same as now).
* To line number: Search `lines` for last entry where `line.utf8_offset <= utf8_offset`. Line number is the index.
* To column number: `(utf8_offset - line.utf8_offset) - (offset.utf16_difference - line.utf16_difference)`.

Despite having 2 tables to do lookups in, I think this is the most efficient design when converting offsets in ascending order. Non-ASCII characters are rare, so the 2nd table will rarely have to be searched. Combining the 2 tables into 1 would require "fatter" rows with more data to cover both line breaks and non-ASCII characters - which would require more memory and be slower to search.

We can construct both tables in a single pass, using SIMD to process chunks of 16 bytes at a time.

Ideally, we'd move construction of the tables into lexer which is already processing the source byte-by-byte, but that's for later.

**Comment by overlookmotel** [comment]
@mertcanaltin is hoping to tackle adding `loc` in next couple of weeks.

Some context and thoughts which I hope may be useful:

## How ASTs get from Rust to JS

We have 2 ways to send an AST from Rust to JS:

1. Via JSON - Serialize to JSON on Rust side, transfer JSON string to JS, then `JSON.parse` on JS side.
2. Experimental "raw transfer" mechanism - send reference to Rust's memory to JS side as `Uint8Array`, deserializer written in JS reads from the buffer and converts to JS objects.

Oxc's Rust AST has mostly the same shape as ESTree, but still there are many differences. Both of these transfer methods also perform translation from Oxc's native AST shape to ESTree shape at the same time as serializing (in method 1) or deserializing (in method 2).

Raw transfer is much faster than JSON, but is experimental and behind an undocumented `experimentalRawTransfer` option, so few people use it.

Eventually, we need to support `range` and `loc` in raw transfer, **but let's ignore that for now**, and just try to make it work for JSON transfer.

## How JSON transfer works

Serialization to JSON is via [`ESTree`](https://github.com/oxc-project/oxc/blob/e327d5e3de4d765243afc0199821bb98fc6ae447/crates/oxc_estree/src/serialize/mod.rs) trait. `ESTree` is much like `serde_json::Serialize`, but specialized for our use case, and about 50% faster than `serde`.

Most types have their `ESTree` implementation auto-generated. These are in [oxc_ast/src/generated/derive_estree.rs](https://github.com/oxc-project/oxc/blob/e327d5e3de4d765243afc0199821bb98fc6ae447/crates/oxc_ast/src/generated/derive_estree.rs), and the codegen is in [tasks/ast_tools/src/derives/estree.rs](https://github.com/oxc-project/oxc/blob/e327d5e3de4d765243afc0199821bb98fc6ae447/tasks/ast_tools/src/derives/estree.rs). The codegen uses various attributes on the AST types themselves to guide it e.g. `#[estree(skip)]`.

Hand-written serializers for types which need custom code to massage them into ESTree shape are in [oxc_ast/src/serialize](https://github.com/oxc-project/oxc/tree/e327d5e3de4d765243afc0199821bb98fc6ae447/crates/oxc_ast/src/serialize).

The complications around AST shape are not relevant to this issue - this info is just for background.

## Offset conversion

Rust strings are UTF-8, whereas JS uses UTF-16. So offsets (`start`, `end`) need to be translated from UTF-8 byte offsets to UTF-16 char offsets in the JS-side AST.

Before the AST is serialized by `ESTree`, there's an initial pass through the AST to convert offsets. [`Utf8ToUtf16`](https://github.com/oxc-project/oxc/blob/e327d5e3de4d765243afc0199821bb98fc6ae447/crates/oxc_ast_visit/src/utf8_to_utf16/mod.rs) handles this. `Utf8ToUtf16::new` builds a translation table, and `Utf8ToUtf16::convert_program` traverses the whole AST, mutating the `Span` of each node to change the offsets from UTF-8 to UTF-16.

There are 2 keys to making this process fast:

1. Generating the translation table involves searching the source text for non-ASCII bytes. Do that in chunks of 32 bytes, which compiler can boil down to a handful of SIMD ops.
2. `convert_program` visits offsets in ascending order. So converting each offset doesn't need to search the whole translation table, but instead advances a cursor through the translation table linearly.

[`Utf8ToUtf16Converter`](https://github.com/oxc-project/oxc/blob/e327d5e3de4d765243afc0199821bb98fc6ae447/crates/oxc_ast_visit/src/generated/utf8_to_utf16_converter.rs) is the AST visitor that performs this in-order traversal.

`Utf8ToUtf16Converter` is generated by [this codegen](https://github.com/oxc-project/oxc/blob/e327d5e3de4d765243afc0199821bb98fc6ae447/tasks/ast_tools/src/generators/utf8_to_utf16.rs). AST types have their fields in source code order, so mostly it's simple to visit in order, but there a few [exceptions](https://github.com/oxc-project/oxc/blob/e327d5e3de4d765243afc0199821bb98fc6ae447/tasks/ast_tools/src/generators/utf8_to_utf16.rs#L30-L47) for which there's [hand-written visitors](https://github.com/oxc-project/oxc/blob/e327d5e3de4d765243afc0199821bb98fc6ae447/crates/oxc_ast_visit/src/utf8_to_utf16/visit.rs).

## How to implement `loc`

There's a few parts to this:

### 1. Expand the translation table and its builder

We need the UTF8 -> UTF16 translation table to include a mapping from offset to line number. See my previous comment on this issue for a design I think would work. But it can be simplified a bit - see (3) below.

It'd be ideal to retain the SIMD-acceleration in building the table.

Line breaks we need to handle are `\r`, `\n`, `\r\n` (considered as 1 line break), `LS`, and `PS`. `LS` and `PS` are Unicode chars, so can be handled in the cold path for non-ASCII. Both have `0xE2` as their first byte.

There's [some similar code in `oxc_codegen`](https://github.com/oxc-project/oxc/blob/e327d5e3de4d765243afc0199821bb98fc6ae447/crates/oxc_codegen/src/comment.rs#L16-L94) (though that's not SIMD-optimized).

### 2. Add a `loc` option to ESTree serializers

This option should be added to the `Config` trait, along the same lines as [`range` option](https://github.com/oxc-project/oxc/blob/70f01439315413b9f66f3851a1e1ac8dbb5bec22/crates/oxc_estree/src/serialize/config.rs).

### 3. Use the translation table in the serializer

The current method for converting `Span`s from UTF-8 to UTF-16 in a pre-pass of AST won't work for `loc`. There's nowhere in AST types to store line/column info. We don't want to add fields to `Span`, because every AST node has a `Span`, so making `Span` bigger would greatly increase the AST's memory usage.

Instead, we'll need to do offset to line+column conversion during serialization.

`serde` doesn't allow state to be stored in the serializer. But as we own `ESTree`, we can adapt it to do that. We would store a reference to the translation table in the serializer.

Then [`serialize_span`](https://github.com/oxc-project/oxc/blob/8b1e022f7fe21e1e175398d0479e28cb385fc93e/crates/oxc_estree/src/serialize/structs.rs#L140-L151) can check if `loc` option is enabled and, if so, use the translation table to convert `start` + `end` offsets to line/column, and output them as `loc`.

This won't be as fast as offset conversion, because we won't be converting offsets in ascending order. But it shouldn't be *too* bad for perf. Trying to do conversion in ascending order would be *way* more complicated, and I suggest leaving that optimization until later.

Important note: The offsets in `Span` will have already been converted in the pre-pass from UTF-8 to UTF-16. So the conversion from offset to line+column is starting from *UTF-16* offset, not UTF-8 offset. This will affect the design of the translation table - it can likely be simplified from what I suggested above.

### 4. Tests

We have a large conformance test suite which checks that JSON-serialized AST matches Acorn for all Test262 test cases.

Those conformance tests cover `start` and `end`. They do not currently cover `range`, because the same test fixtures are used for raw transfer tests too, and `range` is not supported in raw transfer yet. For the same reasons, we also won't be able to test `loc` in the conformance suite.

So we'll need to manually add some tests, alongside [the tests for `range`](https://github.com/oxc-project/oxc/blob/8b1e022f7fe21e1e175398d0479e28cb385fc93e/napi/parser/test/parse.test.ts#L692-L714).

## Final thoughts

Having written this up, I realize it is no small undertaking! I hope the notes above are a help, and @mertcanaltin (or someone else if he's unavailable) may be able to make some progress on this.

**Comment by overlookmotel** [comment]
#14321 adds support for `range` field to raw transfer. Support for `loc` remains a work in progress (#13285).
